lib/inputs/dovetail-frequency.js

```diff
@@ -0,0 +1,159 @@
+      err.skipLogging = true;
+      throw err;
+    }
+
+    return output;
+  }
+
+  // spin up N workers, and process the queue of updates
+  async updateAll(updates, concurrency = 25) {
+    let successCount = 0;
+    let deleteCount = 0;
+    let failureCount = 0;
+
+    const client = await dynamo.client();
+    const worker = async () => {
+      let args;
+      while ((args = updates.shift())) {
+        try {
+          const params = this.updateParams(args);
+          const result = await dynamo.updateItemPromise(params, client);
+
+          const toDelete = this.evictImpressions(result);
+          if (toDelete.length > 0) {
+            const deleteParams = this.deleteParams(args, toDelete);
+            const deleteResult = await dynamo.updateItemPromise(deleteParams, client);
+            deleteCount++;
+          }
+
+          successCount++;
+        } catch (err) {
+          if (err.name === 'ProvisionedThroughputExceededException') {
+            logger.warn(`DDB throughput exceeded [${process.env.DDB_TABLE}]: ${err}`, { args });
+          } else {
+            logger.error(`DDB Error [${process.env.DDB_TABLE}]: ${err}`, { args });
+          }
+          failureCount++;
+        }
+      }
+    };
+
+    const threads = Array(concurrency).fill(true);
+    await Promise.all(threads.map(() => worker()));
+    return [successCount, deleteCount, failureCount];
+  }
+
+  evictImpressions(result) {
+    if (result.Attributes && result.Attributes.impressions) {
+      const evict = Math.floor((new Date().getTime() / 1000) - FREQUENCY_HORIZON_SECONDS);
+      return result.Attributes.impressions.NS.filter((imp) => imp <= evict);
+    }
+    return [];
+  }
+
+  deleteParams(data, deleteArray) {
+    return {
+      TableName: this.tableName(),
+      Key: {
+        listener: { S: `${data.listener}` },
+        campaign: { S: `${data.campaign}` },
+      },
+      ReturnValues: 'ALL_NEW',
+      UpdateExpression: "DELETE impressions :td",
+      ExpressionAttributeValues: {
+        ":td": { "NS": deleteArray.map((td) => `${td}`) },
+      },
+    };
+  }
+
+  updateParams(data) {
+    return {
+      TableName: this.tableName(),
+      Key: {
+        listener: { S: `${data.listener}` },
+        campaign: { S: `${data.campaign}` },
+      },
+      ReturnValues: 'ALL_OLD',
+      UpdateExpression: "SET expiration = :ttl ADD impressions :ts",
+      ExpressionAttributeValues: {
+        ":ttl": { "N": `${(data.timestamp + FREQUENCY_HORIZON_SECONDS)}` },
+        ":ts": { "NS": [`${data.timestamp}`] },
+      },
+    };
+  }
+  
+  async format(record, imp) {
+    let out = {
+      listener: record.listenerId,
+      campaign: imp.campaignId,
+      timestamp: timestamp.toEpochSeconds(record.timestamp || 0),
+    };
+    return out;
+  }
+
+  async eachImpression(handler) {
+    await Promise.all(
+      this._records.map(async rec => {
+        await Promise.all(
+          rec.impressions.map(async imp => {
```

Comment:
But we'll want to filter these more, right?  Skip the dups, and only process campaigns that have "opted into" frequency capping.  Maybe pass some config about the type of frequency cap through to the `handler` method.

Reply:
I think it already skips dupes, but yes, we need to punch through the freqcap info on the impression so we can only bother recording this for the campaigns that are in fact frequency capped.

We'd also at that point have the info about what time frame we need to keep for that campaign, instead of keeping 30 days.

This is very much missing those features, that would need to come through augury and DTR changes.

Reply:
`frequency` now added to the impression log format coming from DTR
https://github.com/PRX/dovetail-router.prx.org/pull/304/files#diff-0190b0822853e6c7bc30be5d2cc8c452c0fe00c53125ac160b20ae11113eaedbR109

So we can now filter for that, and use the values in there to look at what time stamps to evict